File: frontend/src/types/practice.ts
import type { DifficultyLevel, QuestionType } from "./contracts";

export interface PracticeQuestion {
  id: string;
  courseId: string;
  nodeId?: string;
  questionType: QuestionType;
  title: string;
  content: string;
  options?: string[];
  answer: string;
  explanation?: string;
  difficulty: DifficultyLevel;
  tags: string[];
  createdAt: string;
  updatedAt: string;
}

export interface PracticeGenerateRequest {
  userId: string;
  courseId: string;
  nodeId?: string;
  questionTypes: QuestionType[];
  difficulty?: DifficultyLevel;
  count: number;
}

export interface PracticeSubmitRequest {
  userId: string;
  questionId: string;
  userAnswer: string;
  durationSeconds?: number;
}

export interface PracticeRecord {
  id: string;
  userId: string;
  questionId: string;
  nodeId?: string;
  userAnswer: string;
  correctAnswer: string;
  isCorrect: boolean;
  score: number;
  mistakeReason?: string;
  durationSeconds?: number;
  createdAt: string;
  updatedAt: string;
}

File: backend/app/schemas/practice.py
from pydantic import Field

from app.schemas.common import ContractModel, DifficultyLevel, QuestionType


class PracticeQuestion(ContractModel):
    id: str
    course_id: str
    node_id: str | None = None
    question_type: QuestionType
    title: str
    content: str
    options: list[str] | None = None
    answer: str
    explanation: str | None = None
    difficulty: DifficultyLevel
    tags: list[str] = Field(default_factory=list)
    created_at: str
    updated_at: str


class PracticeGenerateRequest(ContractModel):
    user_id: str
    course_id: str
    node_id: str | None = None
    question_types: list[QuestionType]
    difficulty: DifficultyLevel | None = None
    count: int


class PracticeSubmitRequest(ContractModel):
    user_id: str
    question_id: str
    user_answer: str
    duration_seconds: int | None = None


class PracticeRecord(ContractModel):
    id: str
    user_id: str
    question_id: str
    node_id: str | None = None
    user_answer: str
    correct_answer: str
    is_correct: bool
    score: float
    mistake_reason: str | None = None
    duration_seconds: int | None = None
    created_at: str
    updated_at: str
